nifi-nar-bundles/nifi-deeplearning4j-bundle/nifi-deeplearning4j-processors/src/main/java/org/apache/nifi/processors/deeplearning4j/DeepLearning4JPredictor.java

```diff
@@ -0,0 +1,218 @@
+            String message = "FlowFile query is empty";
+            getLogger().error(message);
+            flowFile = session.putAttribute(flowFile, DEEPLEARNING4J_ERROR_MESSAGE, message);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        String input = null;
+        try {
+            input = getFlowFileContents(session, charset, flowFile);
+            String fieldSeparator = context.getProperty(FIELD_SEPARATOR).evaluateAttributeExpressions(flowFile).getValue();
+            String recordSeparator = context.getProperty(RECORD_SEPARATOR).evaluateAttributeExpressions(flowFile).getValue();
+
+            int [] dimensions = getInputDimensions(context, charset, flowFile, fieldSeparator);
+
+            if ( getLogger().isDebugEnabled() )    {
+                getLogger().debug("Received input {} with dimensions {}", new Object[] { input, dimensions });
+            }
+
+            MultiLayerNetwork model = getModel(context);
+
+            long startTimeMillis = System.currentTimeMillis();
+
+            String [] inputRecords = input.split(recordSeparator);
+
+            List<INDArray> features = Arrays.stream(inputRecords).map(
+                record -> {
+                    double [] parameters = Arrays.stream(record.split(fieldSeparator)).mapToDouble(
+                             field -> Double.parseDouble(field)).toArray();
+
+                    INDArray featureInput = Nd4j.create(parameters, dimensions);
+
+                    if ( getLogger().isDebugEnabled() ) {
+                        getLogger().debug("Features for record {} parameters {} dims {} featureInput {} ",
+                            new Object[] {record, parameters, dimensions, featureInput});
+                    }
+
+                    return featureInput;
+
+                }).collect(Collectors.toList());
+
+           INDArray allFeatures = Nd4j.vstack(features);
+
+           INDArray results = model.output(allFeatures);
+
+           double [][] partitionedResults = new double[inputRecords.length][];
+           for (int row = 0; row < inputRecords.length; row++) {
+                INDArray result = results.getRow(row);
+                partitionedResults[row] = Nd4j.toFlattened(result).toDoubleVector();
+           }
+
+           String jsonResult = gson.toJson(partitionedResults);
+           int [] shape = results.shape();
+           String jsonShape = gson.toJson(Arrays.copyOfRange(shape, 1, shape.length));
+
+           if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Prediction for inputRecords {}, dims {}, results {}, result.shape {}, partitionedResults {}, jsonResult {}, shape {}, jsonShape {}",
+                        new Object[] {inputRecords, dimensions, results, Arrays.toString(results.shape()), partitionedResults, jsonResult, shape, jsonShape});
+           }
+
+           try (ByteArrayInputStream bais = new ByteArrayInputStream(jsonResult.getBytes(charset))) {
```

Comment:
Is more efficient here (and probably more straight-forward?) to simply do:
`flowFile = session.write(flowFile, out -> out.write(jsonResult.getBytes(charset)));`

Reply:
Corrected.  Thanks